k2/csrc/fsa_algo.cu

```diff
@@ -247,4 +247,11 @@ void ArcSort(Fsa *fsa) {
   SortSublists<Arc, ArcComparer>(fsa);
 }
 
+void ArcSort(Fsa &src, Fsa *dest, Array1<int32_t> *arc_map /*= nullptr*/) {
+  K2_CHECK_GE(src.NumAxes(), 2) << "Input FSA is empty!";
```

Comment:
Empty fsa is with NumAxes ==2 and Dim0 = 0, this algorthims should work correctly for empty FSA (and seems your implementation works well for empty FSA). I think we should update the log info here to `Invalid FSA`, or just remove the text.

Reply:
`Array1<T>::Clone` will crash for empty arrays in that their region is NULL and `Context()` will segfault.

If the input FSA is empty, what will we put in `dest` and `arc_map`?

Reply:
Empty array is different from Invalid array, you can have a check of method `IsValid` in Array1.

> If the input FSA is empty, what will we put in dest and arc_map
Just leave them empty as well. I think your implementation of SortSublists work well for empty cases.

Reply:
It's possible for arrays to have zero elements but have a non-NULL region.  You can allocate a region of zero bytes, I believe.
Instead of saying "is empty", you can just say "has X axes" or change it to K2_CHECK_GE(x.NumAxes(), 2);

We don't need to waste too many cycles thinking about this though.  Although I try to make these things, where possible, not crash for empty input, in many cases it won't work for this or that reason.  And I think expecting things to work when we provide empty inputs from Python is maybe unreasonable, and in any case not a short-term priority.